blockprotocol/blockprotocol
# Require `$schema` in ontology types

## 🌟 What is the purpose of this PR?

Unfortunately there was a typo where we updated the meta-schemas for the types to have the `schema` keyword instead of the `$schema` keyword. More importantly, there was an oversight and we didn't update the `type-system` definition to handle or enforce this.

This PR updates as necessary.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1203358502199087/1204064995655460/f) _(internal)_

## 🚫 Blocked by

N/A

## 🔍 What does this change?

- Fixes the typos in the meta-schemas
- Updates the `type-system` package to enforce `$schema` to be set to the correct values
- Updates the API implementation to add `$schema` when creating types

## 📜 Does this require a change to the docs?

- The meta-schemas are part of the docs

## ⚠️ Known issues

- Any types currently deployed will need manually patching, we shall opt to do this while the published packages are still canaries. In the future this would need to be a proper update
- Our `ts-node` loader does _not_ like the `@blockprotocol/core` package. Importing it (even indirectly through `@blockprotocol/graph` causes strange errors complaining about CommonJS and named exports. it also causes problems within the site middleware:
  - [Patch broken seed-db](https://github.com/blockprotocol/blockprotocol/pull/1053/commits/3206ccbbf139c0c8bbe893300d2a639f239b28ad)
  - [Patch broken middleware](https://github.com/blockprotocol/blockprotocol/pull/1053/commits/629431153b0dd1ce4dd78cd149cff81619156c8e)
  - [Patch broken middleware again?](https://github.com/blockprotocol/blockprotocol/pull/1053/commits/0ec012af2770ee748d706b5fae124111862fa786)

## 🐾 Next steps

- Update the type editor
- Patch the DB
- Remove temporary patch on type-editor boundary

## 🛡 What tests cover this?

- Type System crate tests
- Type system node package tests

## ❓ How to test this?

1.  Checkout the branch / view the deployment
1.  `yarn` should pass, rebuilding the Type System crate
1.  `yarn workspace @blockprotocol/type-system test` should pass
1.  `cargo make test` should pass inside the `type-system/crate` directory 
1.  `cargo make lint` should pass inside the `type-system/crate` directory
1. Run the site and ensure the type-editor continues to function, and creates types with the correct fields
